python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_callback.py

```diff
@@ -0,0 +1,708 @@
+
+    def end_trace(self, trace_id: Optional[str] = None, *args: Any, **kwargs: Any) -> None:
+        roots, adjacency_list = _build_graph(self._event_data)
+        self._event_data.clear()
+        dfs_stack = roots.copy()
+        while dfs_stack:
+            event_id, event_data = dfs_stack.pop()
+            event_type = event_data.event_type
+            attributes = event_data.attributes
+
+            if event_type is CBEventType.LLM:
+                # Need to update the template variables attribute by extracting their values
+                # from the template events, which are sibling events preceding the LLM event.
+                # Note that the stack should be sorted in chronological order, and we are
+                # going backwards in time on the stack.
+                while dfs_stack:
+                    _, preceding_event_data = dfs_stack[-1]
+                    if (
+                        preceding_event_data.parent_id != event_data.parent_id
+                        or preceding_event_data.event_type is not CBEventType.TEMPLATING
+                    ):
+                        break
+                    dfs_stack.pop()  # remove template event from stack
+                    for payload in preceding_event_data.payloads:
+                        # Add template attributes to the LLM span to which they belong.
+                        try:
+                            attributes.update(_template_attributes(payload))
+                        except Exception:
+                            logger.exception(
+                                f"Failed to convert payload to semantic attributes. "
+                                f"event_type={preceding_event_data.event_type}, payload={payload}",
+                            )
+
+            if event_data.end_time is None:
+                self._stragglers[event_id] = event_data
+            elif not event_data.is_dispatched:
+                _finish_tracing(event_data)
+            dfs_stack.extend(adjacency_list[event_id])
+
+
+class _ContextApiTokenStack:
+    """
+    Allows popping a token from the middle of the stack, whereby all tokens above
+    it are also popped. Popping a non-existent token is allowed and has no effect.
+    """
+
+    __slots__ = ("_stack", "_positions")
+
+    def __init__(self) -> None:
+        self._stack: List[object] = []
+        self._positions: Dict[int, int] = {}
+
+    def append(self, token: object) -> None:
+        if id(token) in self._positions:
+            return
+        position = len(self._stack)
+        self._positions[id(token)] = position
+        self._stack.append(token)
+
+    def pop(self, token: object) -> None:
+        if (position := self._positions.get(id(token))) is None:
+            return
+        for token in reversed(self._stack[position:]):
+            self._positions.pop(id(token))
+            context_api.detach(token)
+        self._stack = self._stack[:position]
```

Comment:
The `self._stack` seems extraneous here. I think you can implement both `append` and `pop` with just `self._positions`.

Reply:
The stack provides ordering, otherwise we don't know which tokens are on top of the current one unless we iterate over the whole dictionary, and even if we do that we still have to put them into a list so they can be detached in the right order.

Reply:
I see. You could also use an OrderedDict.

Reply:
I see. I think you could use an `OrderedDict` and do away with the stack. Small nit.